presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/avro/AvroRowEncoder.java

```diff
@@ -0,0 +1,161 @@
+
+    public AvroRowEncoder(ConnectorSession session, List<EncoderColumnHandle> columnHandles, Schema parsedSchema)
+    {
+        super(session, columnHandles);
+        for (EncoderColumnHandle columnHandle : this.columnHandles) {
+            checkArgument(columnHandle.getFormatHint() == null, "Unexpected format hint '%s' defined for column '%s'", columnHandle.getFormatHint(), columnHandle.getName());
+            checkArgument(columnHandle.getDataFormat() == null, "Unexpected data format '%s' defined for column '%s'", columnHandle.getDataFormat(), columnHandle.getName());
+
+            checkArgument(isSupportedType(columnHandle.getType()), "Unsupported column type '%s' for column '%s'", columnHandle.getType(), columnHandle.getName());
+        }
+        this.byteArrayOuts = new ByteArrayOutputStream();
+        this.parsedSchema = requireNonNull(parsedSchema, "parsedSchema is null");
+        this.dataFileWriter = new DataFileWriter<>(new GenericDatumWriter<>(this.parsedSchema));
+        this.record = new GenericData.Record(this.parsedSchema);
+    }
+
+    private boolean isSupportedType(Type type)
+    {
+        return isVarcharType(type) || SUPPORTED_PRIMITIVE_TYPES.contains(type);
+    }
+
+    @Override
+    protected void appendNullValue()
+    {
+        record.put(columnHandles.get(currentColumnIndex).getName(), null);
+    }
+
+    @Override
+    protected void appendLong(long value)
+    {
+        record.put(columnHandles.get(currentColumnIndex).getName(), value);
+    }
+
+    @Override
+    protected void appendInt(int value)
+    {
+        record.put(columnHandles.get(currentColumnIndex).getName(), value);
+    }
+
+    @Override
+    protected void appendShort(short value)
+    {
+        record.put(columnHandles.get(currentColumnIndex).getName(), value);
+    }
+
+    @Override
+    protected void appendByte(byte value)
+    {
+        record.put(columnHandles.get(currentColumnIndex).getName(), value);
+    }
+
+    @Override
+    protected void appendDouble(double value)
+    {
+        record.put(columnHandles.get(currentColumnIndex).getName(), value);
+    }
+
+    @Override
+    protected void appendFloat(float value)
+    {
+        record.put(columnHandles.get(currentColumnIndex).getName(), value);
+    }
+
+    @Override
+    protected void appendBoolean(boolean value)
+    {
+        record.put(columnHandles.get(currentColumnIndex).getName(), value);
+    }
+
+    @Override
+    protected void appendString(String value)
+    {
+        record.put(columnHandles.get(currentColumnIndex).getName(), value);
+    }
+
+    @Override
+    public byte[] toByteArray()
+    {
+        // make sure entire row has been updated with new values
+        checkArgument(currentColumnIndex == columnHandles.size(), format("Missing %d columns", columnHandles.size() - currentColumnIndex + 1));
+
+        try {
+            byteArrayOuts.reset();
+            dataFileWriter.create(parsedSchema, byteArrayOuts);
```

Comment:
do we need to call this one for each row?
And if should we ignore `DataFileWriter` it returns?

Reply:
This is not a pretty API.  It looks like this call gets the writer ready to write to a new output stream with a new schema.  Feels like a factory method, but it just resets the instance.

And without the schema registry integration, writing Avro messages are quite heavyweight (each message contains the schema).

Reply:
It doesn't work without the `create` method call